Make `remove_linenums!` more precise: remove lines from full Expr (was missing macrocalls), but _NOT_ from user's code (by preventing recursion into `esc()`)

NHDaly opened the pull request:
Fix the `@inferred` and `@test` macros to stop modifying user inputs.

------

After several edits, this PR achieves that by making `Base.remove_linenums!` not recurse into `esc()`'d Exprs.

This PR also merges in https://github.com/JuliaLang/julia/pull/29619, which strengthens `remove_linenums!` to also remove linenumbers from macro calls.

Finally, it introduces a new comparison function, `Base. isequal_nolines(a, b)`, which strips lineinfo nodes from both ASTs and then compares them for equality.

Fixes https://github.com/JuliaLang/julia/issues/31334

ararslan committed:
Remove line info from macro calls in remove_linenums!

Currently `Base.remove_linenums!` removes line information from code
blocks, but not from macro calls, which embed line information as the
second argument in the expression. Having the line information in there
makes it difficult to test user code that performs transformations on
expressions, since the line information may differ in the input and
resulting expressions.

This change extends `Base.remove_linenums!` to handle macro call
expressions, replacing their line information with `nothing`. The result
is:

```julia
julia> :(@macro argument)
:(#= REPL[10]:1 =# @macro argument)

julia> Base.remove_linenums!(:(@macro argument))
:(@macro argument)
```

NHDaly commented:
@vtjnash's suggestion is probably better than what I did here.

... that said, I _have_ always wanted something like this function for recursively replacing part of an AST that matches a pattern. But it's probably not needed for this problem.

NHDaly committed:
Add failing tests showing the broken behavior for `@test` and `@inferred`

NHDaly committed:
Fix at-inferred to not modify inputs

NHDaly committed:
Wrap tests in Testset

NHDaly committed:
Add another test for at-inferred w/ kwargs

NHDaly committed:
try switching to `replace_subtree!` implementation; fixes simple case, unclear how to fix kwargs case

NHDaly committed:
Revert src changes in this PR; tests are all broken again

NHDaly committed:
Make `remove_linenums!` not recurse into `esc()`'d Exprs

This makes sense because the code we introduced that we want to strip
linenumbers from and the user's code we want to `esc()` are
complementary.

 - The reason we use `Base.remove_linenums!` is to avoid injecting the linenumbers of the Test code into the user's test expressions. So we want to remove linenums from any code _iff_ it was introduced by _us_ and not the user.
 - Conversely, we want to always `esc` _any code provided by the user_ exactly once, and we _don't_ want to `esc()` our code, since we want our variable names to get gensym'd.

NHDaly commented:
The trouble now is that all sorts of tests that compare `Expr`s will fail, since `@test` used to strip linenumbers and now it doesn't. Example:
https://github.com/JuliaLang/julia/blob/445f2164a54a6bf28e891180174581d577fc6672/test/syntax.jl#L46-L47

```julia
Test Failed at /Users/sabae/buildbot/worker-tabularasa/tester_macos64/build/share/julia/test/syntax.jl:46
  Expression: $(Expr(:quote, :(module $(Expr(:$, :a))
  #= /Users/sabae/buildbot/worker-tabularasa/tester_macos64/build/share/julia/test/syntax.jl:46 =#
  #= /Users/sabae/buildbot/worker-tabularasa/tester_macos64/build/share/julia/test/syntax.jl:46 =#
  end))) == $(Expr(:quote, :(module a
  #= /Users/sabae/buildbot/worker-tabularasa/tester_macos64/build/share/julia/test/syntax.jl:46 =#
  #= /Users/sabae/buildbot/worker-tabularasa/tester_macos64/build/share/julia/test/syntax.jl:47 =#
  end)))
```

Can we make this kind of change and just require callers to call `Base.remove_linenumbers!` explicitly on their expressions before testing them? (Or maybe we can make a (shorter-named?) function that doesn't modify inputs. Maybe `strip_linenumbers` or `without_linenumbers` or something?)


It's a breaking change, but only breaking tests, which is maybe easier? It's maybe the kind of breaking where no one realllly minds, and we can just change it? (that is, I'm not sure if we actually promised not to change ASTs, and this feels kinda similar?)

The other option seems to me to somehow either change the definition of `==(::Expr, ::Expr)` to remove linenums from both before comparing (Yikes!), or add another comparison function like `isapprox(::Expr, ::Expr)` that does that.

KristofferC commented:
> or add another comparison function like isapprox(::Expr, ::Expr) that does that.

FWIW, I like this.

NHDaly committed:
Add Test.isequal_nolines: remove_linenums! then isequal.

Adds and exports `isequal_nolines` to `Test` stdlib.

NHDaly committed:
Un-export `Test.isequal_nolines`: no reason to export it, the qualified name seems just fine.

NHDaly committed:
Use Test.isequal_nolines to fix julia tests comparing Exprs for equality

NHDaly committed:
Fix `isequal_nolines` docstring: remove line nums from docstring

NHDaly committed:
Merge branch 'aa/aggressive-remove_linenums' into inferred_dont_modify_input

NHDaly committed:
Merge branch 'master' into inferred_dont_modify_input

NHDaly committed:
Change @ararslan's remove-lines to replace w/ `none:1`

This will protect user macros that expect to find a LineNumberNode in
the macrocall AST

DilumAluthge commented:
Honestly, I like the idea of just defining `isapprox(::Expr, ::Expr)`.

You can add a new function called something like `compare_expr`.

But if you use `isapprox`, you can use the Unicode `≈`.

NHDaly commented:
Thanks Dilum! :)

Actually, after thinking about it more, this function should be for parsing _ASTs_, which might be expressions or might be literal values, so we should use a function that could mean something else in the case of a literal.

(Note that ` :(5.0) === 5.0 `, so we should avoid something like `isapprox(:(5.0), :(5.0001))`, which presumably the user _would_ want to find out about.)

So i think this should be its own, Expr-tree-/AST- specific function.

I pushed up some commits that implements this and calls it `Test.isequal_nolines`, which seems explicit/clear enough. I'm also fine with `compare_expr` or `compare_expr_nolines` or `compare_ast_nolines`? I kind of like reusing the `is-blank` format that `isequal` and `isapprox` follow, but i'm happy with whatever everyone can agree on. :)

DilumAluthge commented:
`isequal_nolines` seems fine to me. I like `compare_expr` or `compare_ast` a little bit better because they are shorter. But you make a good point about `isequal_nolines` being consist with `isequal` and `isapprox`.

You could also do `isequal_ast`, which is shorter and maintans the consistency.

NHDaly commented:
(I have merged in https://github.com/JuliaLang/julia/pull/29619, as well, which makes `remove_linenums!` also strip linenumbers from macrocalls, which was missing before, by replacing the line info node with `#= none:1 =#`.)

NHDaly commented:
Please take a look, I think this PR is now ready to be reviewed.

(The existing test failure appears to be a flake.)

Thanks! :)

JeffBezanson commented on stdlib/Test/src/Test.jl:
```diff
@@ -1389,6 +1389,28 @@ function _inferred(ex, mod, allow = :(Union{}))
+See also: [`isapprox`](@ref) for _numerical_ approximate equality
+"""
+function isequal_nolines(a, b)
+    Base.remove_linenums!(a) == Base.remove_linenums!(b)
```
This should `copy` its arguments so the predicate overall is non-mutating.

NHDaly replied:
Oops, duh, thanks.

`deepcopy()` or `copy()`? I feel like `deepcopy`, but i still don't _really_ understand the difference b/w them.

JeffBezanson replied:
`copy` on a container makes a new container with the same elements. So copy of an array allocates exactly one new array. Exprs are more of a tree, so it copies the tree structure (Exprs and all sub-Exprs) but keeps the "leaves". So I think `copy` is correct here.

`deepcopy` attempts to copy everything reachable from an object, as if you serialized the whole thing to disk and read it back in. `deepcopy` is best avoided, since it violates abstraction.

vtjnash replied:
```suggestion
    a = Base.remove_linenums!(copy(a))
    b = Base.remove_linenums!(copy(b))
    return a == b
```

JeffBezanson commented:
I don't think excluding `:escape` expressions is fully correct, it just happens to work in this case. `esc` is just a scope marker, and it's likely we'll change how hygiene works in the future not to use it.

Did you try simply removing the `remove_linenums!` calls from Test and seeing if there are any ill effects? I'm hoping that will work now.

StefanKarpinski commented:
Bump.

vtjnash commented on base/expr.jl:
```diff
@@ -366,9 +366,17 @@ function remove_linenums!(ex::Expr)
+        # Replace line information embedded into macro calls with `nothing`
+        # Removing the argument entirely invalidates the expression
+        map!(ex.args, ex.args) do arg
+            arg isa LineNumberNode ? LineNumberNode(1, :none) : remove_linenums!(arg)
```
These seems controversial, since these are values, not line numbers it’s removing. We could strip the one from the first argument (replacing `__source__`  with `nothing`), but I don’t think we should modify the other argument literals in the list.

>

devmotion replied:
Maybe one could use `LineNumberNode(0, nothing)` as an replacement, to be consistent with the behaviour of `@__LINE__` and `@__FILE__`?

MilesCranmer commented:
Friendly ping

NHDaly commented:
I'm going to close this PR. It is way out of date, and i don't even know if it is the right approach anymore, per Jeff's last message.

Apologies for leaving it dangling. I would love if someone else wants to pick this up! ❤️

vtjnash commented:
Yep, we've eliminated most calls to `remove_linenums!` instead (except from `@inferred` which hasn't been fixed yet to remove that)
